src/types/decimal.rs

```diff
@@ -81,22 +99,28 @@ impl Decimal {
             return d1.exponent.cmp(&d2.exponent);
         }
 
-        // Scale up the magnitude associated with a lesser exponent and compare it with the
-        // other magnitude.
-        if d1.exponent < d2.exponent {
-            Self::compare_scaled_magnitudes(d1, d2)
+        // Scale and compare the coefficients
+        if d1.exponent > d2.exponent {
+            Self::compare_scaled_coefficients(d1, d2)
         } else {
-            Self::compare_scaled_magnitudes(d2, d1)
+            Self::compare_scaled_coefficients(d2, d1).reverse()
         }
     }
 
-    // d1 must have a smaller exponent than d2
-    fn compare_scaled_magnitudes(d1: &Decimal, d2: &Decimal) -> Ordering {
-        let exponent_delta = d2.exponent - d1.exponent;
-        let mut adjusted_magnitude: BigUint = d2.coefficient.magnitude().to_biguint().unwrap();
-        adjusted_magnitude *= 10u64.pow(exponent_delta as u32);
-        let cmp = Magnitude::BigUInt(adjusted_magnitude).cmp(&d1.coefficient.magnitude());
-        cmp
+    // Scales up the coefficient associated with a greater exponent and compares it with the
+    // other coefficient. `d1` must have a larger exponent than `d2`.
+    fn compare_scaled_coefficients(d1: &Decimal, d2: &Decimal) -> Ordering {
+        let exponent_delta = d1.exponent - d2.exponent;
+        // d1 has a larger exponent, so scale up its coefficient to match d2's exponent.
+        // For example, when comparing these values of d1 and d2:
+        //     d1 =  8 * 10^3
+        //     d2 = 80 * 10^2
+        // d1 has the larger exponent (3). We need to scale its coefficient up to d2's 10^2 scale.
+        // We do this by multiplying it times 10^exponent_delta, which is 1 in this case.
+        // This lets us compare 80 and 80, determining that the decimals are equal.
+        let mut scaled_coefficient: BigUint = d1.coefficient.magnitude().to_biguint().unwrap();
+        scaled_coefficient *= 10u64.pow(exponent_delta as u32);
+        Magnitude::BigUInt(scaled_coefficient).cmp(&d2.coefficient.magnitude())
```

Comment:
There's not really a logic change here, but I've revised the variable names to be more accurate and added comments to try and explain what's going on.

Reply:
Thanks for the clarification.